proposals/NNNN-unlock-existential-types-for-all-protocols.md

```diff
@@ -0,0 +1,405 @@
+
+
+### Inconsistent Language Semantics 
+
+The compiler offers existential types for all protocols except for ones that reference `Self` in a [non-covariant](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)) position or/and ones that have associated type requirements. However, that is incohesive as `Self` can be thought of as an associated type on its own, because it's can't be known solely from the context of a protocol; thus, giving `Self` and associated types different rules for synthesis of an existential type seems rather unintuitive. Furthermore, current semantics can promt library authors to avoid refining their protocols with other useful protocols in fear of their refined protocol not qualifying for the automatically generated existential type. Consider the following `Animal` protocol:
+
+```swift
+protocol Identifiable {
+
+  associatedtype ID : Hashable
+
+
+  var id: ID { get }
+
+}
+
+
+protocol Animal : Identifiable where ID == String {
+
+  var name: String { get }
+  
+  var speciesName: String { get }
+  
+  
+  var isPet: Bool { get }
+
+}
+
+extension Animal {
+
+  var id: String {
+    name
+  }
+
+}
+```
+
+In the above case, the `Identifiable` inheritence is very useful. Namely, without any cost to the conforming type, `Animnal` can easily refine [`Identifiable`](https://github.com/apple/swift-evolution/blob/main/proposals/0261-identifiable.md) and gain powerful functionality, as a result. For instance, SwiftUI's [`ForEach`](https://developer.apple.com/documentation/swiftui/foreach) view relies on its data source conforming to `Identifiable`, in which case having such a conformance for free on `Animal`-conforming types is very convenient. Lastly, it's also important to note that in the above example the associated type `ID` is fixed to `String`; therefore, the current limitation seems even more unnecessarily restrictive and further exacerbates the confusion surrounding existential types. 
+
+
+## Proposed Solution
+
+We propose to enable synthesis of existential types for all protocols, in spite of their requirements. Thus, compositions which include protocols that could – before this proposal – not be used as types, will along with these protocols get compiler-generated existential types. Thus code similar to the following will become valid.
+
+```swift
+protocol Animal : Identifiable where ID == String {
+
+  ...
+
+}
+
+struct Dog : Animal { ... }
+
+struct Cat : Animal { ... }
+
+
+let myPets: [Animal] = [
+  Dog(named: "Alex"),
+  Cat(named: "Leo")
+] ✅
+
+myPets.first!.name ℹ️
+// Output: Alex
+```
+
+
+## Detailed Design
+
+### Availability of Existential Type
+
+Namely, under the new rules, we propose that a compiler-generated existential be offered for:
+
+1. a protocol, no matter its requirements;
+2. a protocol-with-protocol composition; and 
+3. a protocol-with-class composition.
```

Comment:
A list seems an overkill when we can say "protocols and protocol compositions". `Protocol & Class` are a kind of protocol composition as per [TSPL](https://docs.swift.org/swift-book/LanguageGuide/Protocols.html)

Reply:
Updated to this:

> Namely, under the new rules, we propose that a compiler-generated existential be offered for protocols and protocol compositions no matter their requirements.